### ISSUE TITLE
Which metric to use

### ISSUE THREAD

**Issue by lizgzil** [wellcometrust/reach#175]
These are the latest results. 'Score' is the metric returned if you do verbose=False.
```
Date of evaluation = 2019-07-16-1716

-----How well the scraper finds the references section:-----

Score
0.944

Lenient F1-score (references section exists or not)
0.944

Metrics by provider
Provider                                         msf    nice  unicef  who_iris        all
Number of pdfs included                       15.000  17.000  31.000    32.000  95.000000
Number of pdfs with sections text              7.000   5.000   9.000     5.000  26.000000
Proportion of pdfs with sections text          0.467   0.294   0.290     0.156   0.274000
Lenient F1 score for all sections included     0.911   0.941   0.914     0.990   0.944000
Number of pdfs with a bibliograph section      1.000   0.000   0.000     1.000   2.000000
Number of pdfs with a endnotes section         2.000   0.000   4.000     0.000   6.000000
Number of pdfs with a reference section        4.000   5.000   5.000     4.000  18.000000
Proportion with a bibliograph section          0.067   0.000   0.000     0.031        NaN
Proportion with a endnotes section             0.133   0.000   0.129     0.000        NaN
Proportion with a reference section            0.267   0.294   0.161     0.125        NaN
Lenient F1 score for the bibliograph section   0.933   1.000   1.000     1.000   0.989474
Lenient F1 score for the endnotes section      0.933   1.000   1.000     1.000   0.989474
Lenient F1 score for the reference section     0.867   0.824   0.742     0.969   0.852632

Number of pdfs with sections text
26

Mean normalised Levenshtein distance
0.612

Strict accuracy (micro)
0.0

Lenient accuracy (micro) (normalised Levenshtein < 0.3)
0.308

Mean normalised Levenshtein distance for the bibliograph section
0.506

Strict accuracy for the bibliograph section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the bibliograph section
0.5

Mean normalised Levenshtein distance for the reference section
0.652

Strict accuracy for the reference section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the reference section
0.278

Mean normalised Levenshtein distance for the endnotes section
0.524

Strict accuracy for the endnotes section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the endnotes section
0.333

-----How well the splitter predicted how many references there were:-----

Score
61.4

Number of reference sections in sample
70

Percentage below threshold of 40
61.4

Median difference
20.5

Metrics grouped by source
Source                                   msf  nice  unicef  who_iris
Number of reference sections in sample   9.0  25.0    14.0      22.0
Median difference                        7.3  95.7    27.4       8.7
Percentage below threshold of 40        88.9  20.0    78.6      86.4

-----How well the parser predicted reference component texts:-----

Score
0.403

Number of references in sample
205

Number of non-blank reference components
935

Strict accuracy (micro)
0.403

Lenient accuracy (micro)(normalised Levenshtein < 0.3)
0.817

Mean normalised Levenshtein distance (all categories)
0.138

Metrics per category
                                                Authors   Issue  Journal  Pagination  PubYear    Title  Volume
Number of non-blank reference components        176.000  40.000  154.000      90.000  188.000  202.000  85.000
Strict accuracy                                   0.341   0.639    0.337       0.546    0.293    0.244   0.420
Lenient accuracy(normalised Levenshtein < 0.3)    0.615   0.917    0.839       0.912    0.956    0.610   0.868
Mean normalised Levenshtein distance              0.299   0.074    0.137       0.055    0.036    0.256   0.108

-----How well the matcher matched references:-----

Score
0.942

Number of references in sample
2000

Micro average F1-score
0.942

Micro average recall
0.942

Micro average precision
0.942

Classification report
              precision    recall  f1-score   support

       False       0.90      1.00      0.94      1000
        True       1.00      0.89      0.94      1000

   micro avg       0.94      0.94      0.94      2000
   macro avg       0.95      0.94      0.94      2000
weighted avg       0.95      0.94      0.94      2000


Frequency table of match types
What did we expect?    Match expected  No match expected
What did we find?                                       
Correct match found               886                  0
Incorrect match found               0                  3
No match found                    114                997

```

**Comment by aCampello** [comment]
I re-ran the evaluation code with the readme.MD instructions for the `evaluation-tidy` branch and was able to get the same results. The steps seem quite comprehensible and very useful to understand each part of the pipeline of the algorithm (maybe the metrics for comparing the entire Section detection should be token-based as discussed with @ivyleavedtoadflax yesterday). Out of curiosity, have the documents selected to the evaluation been used for training the models/calculating the tf-idf matrix? I think it would be good if they were cross-validated.

As far as the overarching metric for the system is concerned, I think we could treat this as an [information retrieval](https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)) (IR) problem as follows. Each policy document, say d_1, ..., d_n is related to the query "Which Wellcome references are in d_i" ? The ground-truth answer for that is a list of references (r_{i1}, r_{i2}, ....).  On the other hand, our end-to-end algorithm will also output another list of predicted references (r'_{i1}, ....). With these two lists, we can calculate IR metrics such as precision, recall or (Jaccard index)[https://en.wikipedia.org/wiki/Jaccard_index] (intersection over the union, usually employed in recommender systems/text-ranking algorithms). This would give an indication of the end-to-end performance of the algorithm in terms of "how many correct references we found".

For us to be able to calculate that, we would need some annotated data that contains pairs of pdfs with their corresponding (ground-truth) matched references. I am not sure if we already have such examples. I know @ivyleavedtoadflax has been putting up an annotation task for the pdfs, and from that we could construct a ground-truth dataset.

**Comment by nsorros** [comment]
I was thinking along the same lines but without making it explicitly about Wellcome.

I was thinking to collect data on `policy_document_id, pulbication_id` and compare with the out of the tool for structuring references (everything apart from matching). Then true_positives are when a set appears in two lists, false_positive is when a set appears in the tool but in ground truth and false negative the other way around. With these metrics we can calculate precision, recall, accuracy, f1 etc.

So pretty much the same but not making it only for Wellcome publications.

**Comment by ivyleavedtoadflax** [comment]
I agree with this. I think it gets to the core of what we are trying to do, and also gives a very clear metric to report.

**Comment by lizgzil** [comment]
Outcome of our meeting today.

**Now:**
FIND - one minus a mean normalised distance metric (whether it be Levenshtein, BLEU or ROUGE)
SPLIT - As is
PARSE - one minus a mean normalised distance metric (whether it be Levenshtein, BLEU or ROUGE)
MATCH - As is

**Interim:**
Matt's annotation work should give us an update to the SPLIT metric. With his data we will have what references text is expected for each reference, so don't need to use our number of references similarity metric.
This data also gives us realistic evaluation data for MATCH, rather than the perfect match set we currently have.

**Future:**
We will have a Gold Standard Metric, which will involve creating an evaluation dataset of policy ID - publication ID pairs. Hence we input the policy ID, run the tool, and get out publication ID(s) of the matches, and then calculate one easy F1/accuracy score.  
The publication IDs will be linked to their doi (which we could find from scraping the internet and then manually checking the results), and thus we can get some lovely metadata in which to understand our biased.

This gold standard metric would be great for now, but is also essential for a future in which we have merged the FIND, SPLIT and PARSE steps.

Hope I remembered everything accurately! lev_dist(this comment, our convo) == 0?
